rep-2014.rst

```diff
@@ -0,0 +1,259 @@
+
+- **performance-per-watt**: total amount of work (generally *bandwidth* or *throughput*) that can be delivered for every watt of power consumed.
+
+- **memory**: the amount of short-term data (not to be confused with storage) required while executing a given task.
+
+These metrics can help determine performance characteristics of a robotic system. Of most relevance for robotic systems we often encounter the *real-time* and *determinism* characteristics defined as follows: 
+
+- **real-time**: ability of completing a task's computations while meeting time deadlines
+- **determinism**: that the same external or internal event leads to the same system behavior, with executions in the same order, each time.
+
+
+For example, a robotic system may be able to perform a task in a short amount of time (*low latency*), but it may not be able to do it in *real-time*. In this case, the system would be considered to be *non-real-time* given the time deadlines imposed. On the other hand, a robotic system may be able to perform a task in *real-time*, but it may not be able to do it in a short amount of time. In this case, the system would be considered to be *non-interactive*. Finally, a robotic system may be able to perform a task in real-time and in a short amount of time, but it may consume a lot of *power*. In this case, the system would be considered to be *non-energy-efficient*.
+
+In another example, a robotic system that can perform a task in 1 second with a power consumption of `2W` is twice as fast (*latency*) as another robotic system that can perform the same task in 2 seconds with a power consumption of `0.5W`. However, the second robotic system is twice as efficient as the first one. In this case, the solution that requires less power would be the best option from an energy efficiency perspective (with a higher *performance-per-watt*). Similarly, a robotic system that has a high bandwidth but consumes a lot of energy might not be the best option for a mobile robot that must operate for a long time on a battery.
+
+Therefore, it is important to consider different of these metrics when benchmarking a robotic system. The metrics presented in this REP are intended to be used as a guideline, and should be adapted to the specific needs of a robot.
+
+
+Methodology for benchmarking performance in ROS 2
+=================================================
+
+In this REP, we **recommend adopting a grey-box and non-functional benchmarking approach** to measure performance and allow to evaluate ROS 2 individual nodes as well as complete computational graphs. To realize it in an architecture-neutral, representative, and reproducible manner, we also recommend using the Linux Tracing Toolkit next generation (`LTTng <https://lttng.org/>`_) through the `ros2_tracing` project, which leverages probes already inserted in the ROS 2 core layers and tools to facilitate benchmarking ROS 2 abstractions.
```

Comment:
There is a lot of lead up to this section, which I think is important, but this is the "meat" of the rep. I think it might be helpful to make this section more self contained and explicitly state what is meant by `gray-box` and `non-functional.`

Personally, I don't find these definitions particularly clear:

`"Grey-box performance testing: More application-specific measure which is capable of watching internal states of the system and can measure (probe) certain points in the system, thus generate the measurement data with minimal interference. Requires to instrument the complete application."`

Just to clarify, gray box here means the code is instrumented at compile time with an external tracer as opposed to say, just creating a time object and measuring the elapsed time? 

`Non-functional performance testing: The measurable performance of those aspects that don't belong to the system's functions. In the motion planning example, the latency the planner, the memory consumption, the CPU usage, etc.`

Just to clarify this section, do we mean that tests should be "end-to-end", as in from the time from when a "function" is called until it returns? This would be in contrast to say measuring every sub-routine of a "function" and only evaluating the "meaty" part of the system. 

Since this is a recommendation to the community, many of whom are not experts in this sub-domain, we need to be clear, plain-spoken, and direct with our recommendations. A smart undergrad should be able to read this REP and clearly understand how to approach profiling.